File: recommender.py
"""
Movie Recommendation Engine
Implements content-based filtering using TF-IDF and cosine similarity
"""

import pandas as pd
import numpy as np
from typing import List, Dict, Tuple
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import streamlit as st


class MovieRecommender:
    """
    Content-based movie recommendation system.
    
    Uses weighted features:
    - Plot descriptions (40%): TF-IDF vectorization
    - Genres (30%): Binary overlap
    - Director (15%): Exact match
    - Cast (15%): Member overlap
    """
    
    def __init__(self, movie_data: pd.DataFrame):
        """
        Initialize the recommender with movie data.
        
        Args:
            movie_data: DataFrame with columns: title, overview, genres, director, cast, etc.
        """
        self.movie_data = movie_data.copy()
        self.tfidf_matrix = None
        self.tfidf_vectorizer = None
        
        # Precompute TF-IDF matrix for plot descriptions
        self._build_tfidf_matrix()
    
    def _build_tfidf_matrix(self):
        """Build TF-IDF matrix from movie plot descriptions."""
        # Combine overview with genres for richer text features
        self.movie_data['combined_features'] = (
            self.movie_data['overview'].fillna('') + ' ' +
            self.movie_data['genres'].apply(lambda x: ' '.join(x) if isinstance(x, list) else '')
        )
        
        # Create TF-IDF vectorizer
        self.tfidf_vectorizer = TfidfVectorizer(
            max_features=5000,
            stop_words='english',
            ngram_range=(1, 2),  # Include bigrams for better context
            min_df=2  # Minimum document frequency
        )
        
        # Fit and transform
        self.tfidf_matrix = self.tfidf_vectorizer.fit_transform(
            self.movie_data['combined_features']
        )
    
    def _calculate_plot_similarity(self, movie_indices: List[int]) -> np.ndarray:
        """
        Calculate cosine similarity based on plot descriptions.
        
        Args:
            movie_indices: List of movie indices to compare against
            
        Returns:
            Array of similarity scores for all movies
        """
        # Get TF-IDF vectors for selected movies
        selected_vectors = self.tfidf_matrix[movie_indices]
        
        # Calculate cosine similarity with all movies
        similarities = cosine_similarity(selected_vectors, self.tfidf_matrix)
        
        # Average similarity across selected movies
        avg_similarity = similarities.mean(axis=0)
        
        return avg_similarity
    
    def _calculate_genre_similarity(self, movie_indices: List[int]) -> np.ndarray:
        """
        Calculate genre overlap similarity.
        
        Args:
            movie_indices: List of movie indices to compare against
            
        Returns:
            Array of genre similarity scores (Jaccard similarity)
        """
        similarities = np.zeros(len(self.movie_data))
        
        # Get genres of selected movies
        selected_genres = []
        for idx in movie_indices:
            genres = self.movie_data.iloc[idx]['genres']
            if isinstance(genres, list):
                selected_genres.extend(genres)
        selected_genres = set(selected_genres)
        
        # Calculate Jaccard similarity for each movie
        for pos, (i, row) in enumerate(self.movie_data.iterrows()):
            movie_genres = set(row['genres']) if isinstance(row['genres'], list) else set()
            
            if not movie_genres or not selected_genres:
                similarities[pos] = 0
            else:
                intersection = len(selected_genres.intersection(movie_genres))
                union = len(selected_genres.union(movie_genres))
                similarities[pos] = intersection / union if union > 0 else 0
        
        return similarities
    
    def _calculate_director_similarity(self, movie_indices: List[int]) -> np.ndarray:
        """
        Calculate director match similarity (binary: 1 if match, 0 otherwise).
        
        Args:
            movie_indices: List of movie indices to compare against
            
        Returns:
            Array of director similarity scores
        """
        similarities = np.zeros(len(self.movie_data))
        
        # Get directors of selected movies
        selected_directors = set()
        for idx in movie_indices:
            director = self.movie_data.iloc[idx]['director']
            if pd.notna(director):
                selected_directors.add(director)
        
        # Check each movie's director
        for pos, (i, row) in enumerate(self.movie_data.iterrows()):
            if pd.notna(row['director']) and row['director'] in selected_directors:
                similarities[pos] = 1.0
        
        return similarities
    
    def _calculate_cast_similarity(self, movie_indices: List[int]) -> np.ndarray:
        """
        Calculate cast overlap similarity.
        
        Args:
            movie_indices: List of movie indices to compare against
            
        Returns:
            Array of cast similarity scores
        """
        similarities = np.zeros(len(self.movie_data))
        
        # Get cast members of selected movies
        selected_cast = []
        for idx in movie_indices:
            cast = self.movie_data.iloc[idx]['cast']
            if isinstance(cast, list):
                selected_cast.extend(cast[:5])  # Top 5 cast members
        selected_cast = set(selected_cast)
        
        # Calculate overlap for each movie
        for pos, (i, row) in enumerate(self.movie_data.iterrows()):
            movie_cast = set(row['cast'][:5]) if isinstance(row['cast'], list) else set()
            
            if not movie_cast or not selected_cast:
                similarities[pos] = 0
            else:
                overlap = len(selected_cast.intersection(movie_cast))
                similarities[pos] = overlap / 5.0  # Normalize by max possible overlap
        
        return similarities
    
    def _generate_explanation(self, selected_movies: List[str], 
                              movie_indices: List[int],
                              recommended_movie: pd.Series,
                              similarity_components: Dict[str, float]) -> str:
        """
        Generate a human-readable explanation for why a movie was recommended.
        
        Args:
            selected_movies: List of selected movie titles
            movie_indices: Indices of selected movies in the dataframe
            recommended_movie: Series containing recommended movie data
            similarity_components: Dictionary with component similarities
            
        Returns:
            Explanation string
        """
        explanations = []
        
        # Get recommended movie's attributes
        rec_director = recommended_movie.get('director')
        rec_cast = set(recommended_movie.get('cast', [])[:5]) if isinstance(recommended_movie.get('cast'), list) else set()
        rec_genres = set(recommended_movie.get('genres', [])) if isinstance(recommended_movie.get('genres'), list) else set()
        
        # Track which selected movie matches for each component
        director_match_movie = None
        cast_match_movie = None
        genre_match_movie = None
        best_genre_overlap = []
        
        # Check each selected movie for matches
        for i, idx in enumerate(movie_indices):
            selected_movie_data = self.movie_data.iloc[idx]
            sel_director = selected_movie_data.get('director')
            sel_cast = set(selected_movie_data.get('cast', [])[:5]) if isinstance(selected_movie_data.get('cast'), list) else set()
            sel_genres = set(selected_movie_data.get('genres', [])) if isinstance(selected_movie_data.get('genres'), list) else set()
            
            # Director match
            if rec_director and sel_director and rec_director == sel_director:
                director_match_movie = selected_movies[i]
            
            # Cast overlap
            if rec_cast and sel_cast and len(rec_cast.intersection(sel_cast)) > 0:
                cast_match_movie = selected_movies[i]
            
            # Genre overlap - track best match
            genre_overlap = rec_genres.intersection(sel_genres)
            if len(genre_overlap) > len(best_genre_overlap):
                best_genre_overlap = list(genre_overlap)
                genre_match_movie = selected_movies[i]
        
        # Find strongest component
        strongest = max(similarity_components.items(), key=lambda x: x[1])
        
        # Determine which movie to attribute the match to based on strongest signal
        if strongest[0] == 'director' and director_match_movie:
            matched_movie = director_match_movie
        elif strongest[0] == 'cast' and cast_match_movie:
            matched_movie = cast_match_movie
        elif strongest[0] == 'genre' and genre_match_movie:
            matched_movie = genre_match_movie
        elif director_match_movie:
            matched_movie = director_match_movie
        elif cast_match_movie:
            matched_movie = cast_match_movie
        elif genre_match_movie:
            matched_movie = genre_match_movie
        else:
            matched_movie = selected_movies[0]  # Default to first
        
        # Generate explanation based on signals
        if similarity_components.get('genre', 0) > 0.5 and best_genre_overlap:
            explanations.append(f"Shares genres: {', '.join(best_genre_overlap[:2])}")
        
        if similarity_components.get('director', 0) > 0.9:
            explanations.append(f"Same director: {recommended_movie['director']}")
        
        if similarity_components.get('cast', 0) > 0.2:
            explanations.append("Features similar actors")
        
        if similarity_components.get('plot', 0) > 0.3:
            explanations.append("Similar themes and storytelling")
        
        if not explanations:
            explanations.append("Strong overall match based on multiple factors")
        
        # Use the matched movie in the explanation
        base_text = f"Because you liked {matched_movie}"
        explanation_text = " → " + "; ".join(explanations)
        
        return base_text + explanation_text
    
    def get_recommendations(self, selected_titles: List[str], 
                          n_recommendations: int = 5) -> List[Dict]:
        """
        Generate movie recommendations based on user selections.
        
        Args:
            selected_titles: List of movie titles the user likes
            n_recommendations: Number of recommendations to return
            
        Returns:
            List of recommendation dictionaries with movie info and explanations
        """
        # Get indices of selected movies
        movie_indices = []
        for title in selected_titles:
            matching_movies = self.movie_data[self.movie_data['title'] == title]
            if not matching_movies.empty:
                movie_indices.append(matching_movies.index[0])
        
        if not movie_indices:
            return []
        
        # Calculate component similarities with weights
        plot_sim = self._calculate_plot_similarity(movie_indices) * 0.40
        genre_sim = self._calculate_genre_similarity(movie_indices) * 0.30
        director_sim = self._calculate_director_similarity(movie_indices) * 0.15
        cast_sim = self._calculate_cast_similarity(movie_indices) * 0.15
        
        # Combine into final similarity score
        combined_similarity = plot_sim + genre_sim + director_sim + cast_sim
        
        # Exclude already selected movies
        for idx in movie_indices:
            combined_similarity[idx] = -1
        
        # Get top N recommendations
        top_indices = np.argsort(combined_similarity)[::-1][:n_recommendations]
        
        # Build recommendation list with explanations
        recommendations = []
        for idx in top_indices:
            if combined_similarity[idx] > 0:  # Only include positive matches
                movie = self.movie_data.iloc[idx]
                
                # Store component scores for explanation
                component_scores = {
                    'plot': plot_sim[idx] / 0.40,
                    'genre': genre_sim[idx] / 0.30,
                    'director': director_sim[idx] / 0.15,
                    'cast': cast_sim[idx] / 0.15
                }
                
                recommendations.append({
                    'movie': movie.to_dict(),
                    'similarity_score': combined_similarity[idx],
                    'explanation': self._generate_explanation(
                        selected_titles,
                        movie_indices,
                        movie,
                        component_scores
                    )
                })
        
        return recommendations

File: cinematch-js/recommender.js
/**
 * CineMatch - Recommendation Engine (JavaScript)
 * Implements TF-IDF vectorization and cosine similarity for movie recommendations
 */

class MovieRecommender {
    constructor(movies) {
        this.movies = movies;
        this.vocabulary = new Map();
        this.idf = new Map();
        this.tfidfVectors = [];
        this._buildVocabulary();
        this._calculateIDF();
        this._buildTFIDFVectors();
    }

    /**
     * Tokenize and normalize text
     */
    _tokenize(text) {
        if (!text) return [];
        return text.toLowerCase()
            .replace(/[^\w\s]/g, ' ')
            .split(/\s+/)
            .filter(word => word.length > 2);
    }

    /**
     * Build vocabulary from all movie overviews
     */
    _buildVocabulary() {
        let wordIndex = 0;
        for (const movie of this.movies) {
            const tokens = this._tokenize(movie.overview);
            for (const token of tokens) {
                if (!this.vocabulary.has(token)) {
                    this.vocabulary.set(token, wordIndex++);
                }
            }
        }
    }

    /**
     * Calculate Inverse Document Frequency for each term
     */
    _calculateIDF() {
        const docCount = this.movies.length;
        const termDocCounts = new Map();

        // Count documents containing each term
        for (const movie of this.movies) {
            const tokens = new Set(this._tokenize(movie.overview));
            for (const token of tokens) {
                termDocCounts.set(token, (termDocCounts.get(token) || 0) + 1);
            }
        }

        // Calculate IDF
        for (const [term, count] of termDocCounts) {
            this.idf.set(term, Math.log(docCount / (1 + count)));
        }
    }

    /**
     * Build TF-IDF vectors for all movies
     */
    _buildTFIDFVectors() {
        for (const movie of this.movies) {
            const tokens = this._tokenize(movie.overview);
            const termFreq = new Map();

            // Calculate term frequency
            for (const token of tokens) {
                termFreq.set(token, (termFreq.get(token) || 0) + 1);
            }

            // Build sparse vector
            const vector = new Map();
            for (const [term, freq] of termFreq) {
                const tf = freq / tokens.length;
                const idf = this.idf.get(term) || 0;
                vector.set(term, tf * idf);
            }

            this.tfidfVectors.push(vector);
        }
    }

    /**
     * Calculate cosine similarity between two TF-IDF vectors
     */
    _cosineSimilarity(vec1, vec2) {
        let dotProduct = 0;
        let norm1 = 0;
        let norm2 = 0;

        // Calculate dot product and norms
        for (const [term, val1] of vec1) {
            norm1 += val1 * val1;
            if (vec2.has(term)) {
                dotProduct += val1 * vec2.get(term);
            }
        }

        for (const [, val2] of vec2) {
            norm2 += val2 * val2;
        }

        if (norm1 === 0 || norm2 === 0) return 0;
        return dotProduct / (Math.sqrt(norm1) * Math.sqrt(norm2));
    }

    /**
     * Calculate plot similarity for a set of selected movies
     */
    _calculatePlotSimilarity(selectedIndices) {
        const similarities = new Array(this.movies.length).fill(0);

        for (let i = 0; i < this.movies.length; i++) {
            if (selectedIndices.includes(i)) continue;

            let maxSim = 0;
            for (const selIdx of selectedIndices) {
                const sim = this._cosineSimilarity(this.tfidfVectors[selIdx], this.tfidfVectors[i]);
                maxSim = Math.max(maxSim, sim);
            }
            similarities[i] = maxSim;
        }

        return similarities;
    }

    /**
     * Calculate genre similarity using Jaccard index
     */
    _calculateGenreSimilarity(selectedIndices) {
        const similarities = new Array(this.movies.length).fill(0);

        // Get all genres from selected movies
        const selectedGenres = new Set();
        for (const idx of selectedIndices) {
            const genres = this.movies[idx].genres || [];
            genres.forEach(g => selectedGenres.add(g));
        }

        for (let i = 0; i < this.movies.length; i++) {
            const movieGenres = new Set(this.movies[i].genres || []);

            if (movieGenres.size === 0 || selectedGenres.size === 0) {
                similarities[i] = 0;
            } else {
                const intersection = [...movieGenres].filter(g => selectedGenres.has(g)).length;
                const union = new Set([...movieGenres, ...selectedGenres]).size;
                similarities[i] = union > 0 ? intersection / union : 0;
            }
        }

        return similarities;
    }

    /**
     * Calculate director similarity (binary match)
     */
    _calculateDirectorSimilarity(selectedIndices) {
        const similarities = new Array(this.movies.length).fill(0);

        const selectedDirectors = new Set();
        for (const idx of selectedIndices) {
            if (this.movies[idx].director) {
                selectedDirectors.add(this.movies[idx].director);
            }
        }

        for (let i = 0; i < this.movies.length; i++) {
            if (this.movies[i].director && selectedDirectors.has(this.movies[i].director)) {
                similarities[i] = 1.0;
            }
        }

        return similarities;
    }

    /**
     * Calculate cast similarity based on overlapping actors
     */
    _calculateCastSimilarity(selectedIndices) {
        const similarities = new Array(this.movies.length).fill(0);

        const selectedCast = new Set();
        for (const idx of selectedIndices) {
            const cast = this.movies[idx].cast || [];
            cast.slice(0, 5).forEach(c => selectedCast.add(c));
        }

        for (let i = 0; i < this.movies.length; i++) {
            const movieCast = new Set((this.movies[i].cast || []).slice(0, 5));

            if (movieCast.size === 0 || selectedCast.size === 0) {
                similarities[i] = 0;
            } else {
                const intersection = [...movieCast].filter(c => selectedCast.has(c)).length;
                similarities[i] = intersection / movieCast.size;
            }
        }

        return similarities;
    }

    /**
     * Find which selected movie matches the recommendation
     */
    _findMatchedMovie(selectedTitles, selectedIndices, recMovie) {
        let matchedMovie = selectedTitles[0];

        const recDirector = recMovie.director;
        const recCast = new Set((recMovie.cast || []).slice(0, 5));
        const recGenres = new Set(recMovie.genres || []);

        let bestGenreOverlap = [];

        for (let i = 0; i < selectedIndices.length; i++) {
            const selMovie = this.movies[selectedIndices[i]];
            const selDirector = selMovie.director;
            const selCast = new Set((selMovie.cast || []).slice(0, 5));
            const selGenres = new Set(selMovie.genres || []);

            // Director match (highest priority)
            if (recDirector && selDirector && recDirector === selDirector) {
                return { matched: selectedTitles[i], reason: 'director', detail: recDirector };
            }

            // Cast overlap
            const castOverlap = [...recCast].filter(c => selCast.has(c));
            if (castOverlap.length > 0) {
                matchedMovie = selectedTitles[i];
            }

            // Genre overlap
            const genreOverlap = [...recGenres].filter(g => selGenres.has(g));
            if (genreOverlap.length > bestGenreOverlap.length) {
                bestGenreOverlap = genreOverlap;
                if (castOverlap.length === 0) {
                    matchedMovie = selectedTitles[i];
                }
            }
        }

        return { matched: matchedMovie, genres: bestGenreOverlap };
    }

    /**
     * Generate explanation for recommendation
     */
    _generateExplanation(selectedTitles, selectedIndices, recMovie, scores) {
        const explanations = [];
        const matchInfo = this._findMatchedMovie(selectedTitles, selectedIndices, recMovie);

        if (scores.genre > 0.5 && matchInfo.genres && matchInfo.genres.length > 0) {
            explanations.push(`Shares genres: ${matchInfo.genres.slice(0, 2).join(', ')}`);
        }

        if (scores.director > 0.9) {
            explanations.push(`Same director: ${recMovie.director}`);
        }

        if (scores.cast > 0.2) {
            explanations.push('Features similar actors');
        }

        if (scores.plot > 0.3) {
            explanations.push('Similar themes and storytelling');
        }

        if (explanations.length === 0) {
            explanations.push('Strong overall match based on multiple factors');
        }

        return `Because you liked ${matchInfo.matched} → ${explanations.join('; ')}`;
    }

    /**
     * Get movie recommendations based on selected titles
     */
    getRecommendations(selectedTitles, numRecommendations = 5) {
        // Find indices of selected movies
        const selectedIndices = [];
        for (const title of selectedTitles) {
            const idx = this.movies.findIndex(m => m.title === title);
            if (idx !== -1) selectedIndices.push(idx);
        }

        if (selectedIndices.length === 0) return [];

        // Calculate component similarities
        const plotSim = this._calculatePlotSimilarity(selectedIndices);
        const genreSim = this._calculateGenreSimilarity(selectedIndices);
        const directorSim = this._calculateDirectorSimilarity(selectedIndices);
        const castSim = this._calculateCastSimilarity(selectedIndices);

        // Combine with weights
        const combined = this.movies.map((_, i) => {
            if (selectedIndices.includes(i)) return -1;
            return (plotSim[i] * 0.40) + (genreSim[i] * 0.30) +
                (directorSim[i] * 0.15) + (castSim[i] * 0.15);
        });

        // Get top recommendations
        const indices = combined
            .map((score, i) => ({ score, i }))
            .filter(x => x.score > 0)
            .sort((a, b) => b.score - a.score)
            .slice(0, numRecommendations)
            .map(x => x.i);

        // Build recommendation objects
        const recommendations = indices.map(idx => {
            const movie = this.movies[idx];
            const scores = {
                plot: plotSim[idx] / 0.40,
                genre: genreSim[idx] / 0.30,
                director: directorSim[idx] / 0.15,
                cast: castSim[idx] / 0.15
            };

            return {
                movie,
                score: combined[idx],
                explanation: this._generateExplanation(selectedTitles, selectedIndices, movie, scores)
            };
        });

        return recommendations;
    }
}

// Export for use in app.js
window.MovieRecommender = MovieRecommender;
